compiler/rustc_infer/src/infer/error_reporting/mod.rs

```diff
@@ -1783,14 +1783,19 @@ impl<'tcx> TypeErrCtxt<'_, 'tcx> {
                             }
                         }))
                     {
-                        diag.note_expected_found_extra(
-                            &expected_label,
-                            expected,
-                            &found_label,
-                            found,
-                            &sort_string(values.expected, exp_p),
-                            &sort_string(values.found, found_p),
-                        );
+                        if let Some(ExpectedFound { found: found_ty, .. }) = exp_found {
+                            let is_future = self.tcx.ty_is_opaque_future(found_ty);
+                            if !is_future {
```

Comment:
What is this doing? Please leave a comment that explains why we're not nothing this if it's a future.

----

nit: this is equally as readable.

```suggestion
                            if !self.tcx.ty_is_opaque_future(found_ty) {
```

Reply:
Since the Future is a special opaque type that in most of the time the user will not touch directly, print a mismatch type is not so helpful for some target user.

The example in this PR is
```
 |     async_fn()
  |     ^^^^^^^^^^ expected `u8`, found opaque type
  |
  = note:     expected type `u8`
          found opaque type `impl Future<Output = u8>`
```

Where the Future type is never mentioned in the following code

```
fn foo() -> u8 {
    async fn async_fn() -> u8 {  22 }
    
    async_fn()
}
```

I think this will help to improve the errors with async programming for new users, for this reason this diff is skipping the suggestion if it is a future type